src/WebJobs.Script/Host/WorkerFunctionMetadataProvider.cs

```diff
@@ -7,90 +7,108 @@
         }
 
         public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
+        public async Task<FunctionMetadataResult> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
         {
             IEnumerable<FunctionMetadata> functions = new List<FunctionMetadata>();
             _logger.FunctionMetadataProviderParsingFunctions();
 
             if (_functions.IsDefaultOrEmpty || forceRefresh)
             {
                 IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
-                bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, environment);
 
-                if (workerIndexing)
+                if (_channelManager == null)
                 {
-                    if (_dispatcher == null)
-                    {
-                        throw new InvalidOperationException(nameof(_dispatcher));
-                    }
-
-                    // start up GRPC channels
-                    await _dispatcher.InitializeAsync(new List<FunctionMetadata>());
+                    throw new InvalidOperationException(nameof(_channelManager));
+                }
 
-                    // get function metadata from worker, then validate it
-                    rawFunctions = await _dispatcher.GetWorkerMetadata();
+                // start up GRPC channels
+                await _channelManager.InitializeChannelAsync(_workerRuntime);
 
-                    if (IsDefaultIndexingRequired(rawFunctions))
-                    {
-                        _logger.LogDebug("Fallback to host indexing as worker denied indexing");
-                        functions = await _hostFunctionMetadataProvider.GetFunctionMetadataAsync(workerConfigs, environment, forceRefresh);
-                    }
-                    else if (!IsNullOrEmpty(rawFunctions))
+                var channels = _channelManager.GetChannels(_workerRuntime);
+                foreach (string workerId in channels.Keys.ToList())
+                {
+                    if (channels.TryGetValue(workerId, out TaskCompletionSource<IRpcWorkerChannel> initializedLanguageWorkerChannelTask))
                     {
-                        functions = ValidateMetadata(rawFunctions);
+                        _logger.LogDebug("Found initialized language worker channel for runtime: {workerRuntime} workerId:{workerId}", _workerRuntime, workerId);
+                        try
+                        {
+                            IRpcWorkerChannel channel = await initializedLanguageWorkerChannelTask.Task;
```

Comment:
Since we are starting this channel here, do we need to do anything in this class to dispose the channel properly?

Reply:
it doesn't look like we need to do that. I don't see us doing that in other places where we start a channel. Also, we have events that watch and manage channel lifetime as expected,